include/swift/AST/Decl.h

```diff
@@ -3859,6 +3859,32 @@ struct SelfReferenceKind {
       other(other) { }
 };
 
+/// Used to optimize algorithms that compute support kinds
+/// and provide better diagnostics on the reason an existential
+/// type is not supported. For example, the knowledge that an
+/// inherited protocol is unsupported due to a \c Self reference
+/// saves us from searching for associated types and same-type
+/// constraints.
+enum class ExistentialSupportKind {
```

Comment:
Can you split off a change that introduces this enum into a separate commit? It would be good to have separate diagnostics for this too, so it tells you which one of "Self or associated type requirements" the protocol actually has

Reply:
Sure, I was going to separate improvements to diagnostics later into a separate PR, since there will probably be a lot of test corrections plus new tests, but we can append them here as well.
To clarify, you want me to separate just the enum declaration, or all the code that now depends on it?